fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java

```diff
@@ -315,8 +325,110 @@ public boolean isWritable() {
+            } else {
+                if (statusCode != SC_NO_CONTENT) {
+                    LOGGER.error("error committing back transaction {}: {} {}", uri,
+                            statusCode, status.getReasonPhrase());
+                    throw new FedoraException("error committing transaction " + uri + ": " + statusCode + " " +
+                            status.getReasonPhrase());
+                }
+            }
+            LOGGER.debug("Committed transaction {}", TransactionIdHolder.getCurrentTransactionId());
+        } catch (final Exception e) {
+            LOGGER.error("could not encode URI parameter", e);
+            throw new FedoraException(e);
+        } finally {
+            post.releaseConnection();
+            TransactionIdHolder.removeCurrentTransactionId();
+        }
+    }
+
+    @Override
+    public void rollbackTransaction() throws FedoraException {
+        final HttpPost post = httpHelper.createPostMethod(TransactionIdHolder.getCurrentTransactionId()
+                + FCR_TX + FCR_ROLLBACK, null);
+        try {
+            final HttpResponse response = httpHelper.execute(post);
+            final String uri = post.getURI().toString();
+            final StatusLine status = response.getStatusLine();
+            final int statusCode = status.getStatusCode();
+
+            if (statusCode == SC_FORBIDDEN) {
+                LOGGER.error("request to rollback transaction {} is not authorized.", uri);
+                throw new ForbiddenException("request to rollback transaction " + uri + " is not authorized.");
+            } else {
+                if (statusCode != SC_NO_CONTENT) {
+                    LOGGER.error("error rolling back transaction {}: {} {}", uri, statusCode, status.getReasonPhrase());
+                    throw new FedoraException("error rolling back transaction " + uri + ": " + statusCode + " " +
+                            status.getReasonPhrase());
+                }
+            }
+
+            LOGGER.debug("Rolled back transaction {}", TransactionIdHolder.getCurrentTransactionId());
+        } catch (final Exception e) {
+            LOGGER.error("could not encode URI parameter", e);
+            throw new FedoraException(e);
+        } finally {
+            post.releaseConnection();
+            TransactionIdHolder.removeCurrentTransactionId();
+        }
+    }
+
+    @Override
     public String getRepositoryUrl() {
         return repositoryURL;
     }
 
+    private String prependTransactionId(final String path) {
+        // append tx id to the path if there is a transaction present in txHolder,
+        // check if path needs to be prepended with a slash
+        final String txId = TransactionIdHolder.getCurrentTransactionId();
+        return (txId == null || path.contains(TX))
+                ? path : (path.startsWith("/") ? "/" + txId + path : txId + "/" + path);
```

Comment:
I think it would be good to avoid nested ?: conditions -- they are hard enough to read without being nested.